# Tree Building and Phylogenetic Analysis

# Contents

* [4.2.1b: Tree building](#4.2.1-tree-building)
	+ [4.2.1b.1: Preparation](#4.2.1.1-preparation)
	+ [4.2.1b.2: Alignment](#4.2.1.2-alignment)
	+ [4.2.1b.3: Model selection](#4.2.1.3-model-selection)
	+ [4.2.1b.4: Tree building](#4.2.1.4-tree-building)
	+ [4.2.1b.5: Tree visualisation](#4.2.1.5-tree-visualisation)
	+ [4.2.1b.6: Extracting metadata](#4.2.1.6-extracting-metadata)
	+ [4.2.1b.7: Annotating the tree](#4.2.1.7-annotating-the-tree)
	+ [4.2.1b.8: Additional activity: Extracting alternate metadata](#4.2.1.8-additional-activity-extracting-alternate-metadata)


# 4.2.1b: Tree building

In this practical we will be using Omicron SARS-CoV-2 sequences. Due to time constraints, we will be using a subset of sequences (trees can take a very long time to run if there are a lot of sequences!). A representative set of sequences from multiple countries has been selected.

We will be analysing and interpreting best fitting substitution models, bootstrap values and finding the best ways to visualise trees. We will also be extracting useful metadata from sequence IDs and annotating this on to phylogenetic trees. 

You will be using the terminal for some of this, but navigating to the relevant directory with ```cd``` is the only command you'll need to remember - any other new commands will be detailed here. 

## 4.2.1b.1: Preparation

Start by making a folder for the analysis. Make sure this is somewhere easy to navigate to in the terminal! 

You'll then need to copy the fasta file `omicron_subset_countries.fasta` we'll use for this practical from the `~/Phylogenetics` folder. You can do this manually or practice your command line skills! 

## 4.2.1b.2: Alignment

To make trees, sequences first need to be aligned! We will do this using mafft.

Open up the terminal and navigate to the directory you just made using `cd`. 

Write the command `mafft omicron_subset_countries.fasta > omicron_subset_countries_aligned.fasta` 

This will create the file `omicron_subset_countries_aligned.fasta` in your folder - this contains the sequences from `omicron_subset_countries.fasta` aligned by the FFT-NS-2 algorithm.

It's generally a good idea to take a quick manual look at the alignment file before proceeding. You can open the fasta file in an alignment viewer for a full look, or open in a text editor for a quick look. For now, just open the file in a text editor.

### Task 1
How many sequences does this fasta file contain?

Hint: You can do `ctrl+f` to search the file, and search for the `>` character, of which each will denote a sequence. The number of `>`'s should display in the search bar. 

## 4.2.1b.3: Model selection

Finding the substitution model that best fits the data is very important. There are various tools to estimate the best substitution model but for now, we're going to use the inbuilt modeltest function in IQTREE.

### About the program - IQTREE

This  program  allows  you  to  perform  Maximum  Likelihood phylogenetic  analysis.  It  uses  efficient algorithms  to explore the tree  space,  allowing  very  large  matrices  to  be  analyzed  with  reliable  results (hundreds  /  thousands  of  sequences).  It  allows  estimating  the  evolutionary model (ModelFinder module) followed  by the  phylogenetic analysis and implements  support  measures  to evaluate  the reliability  of  the  groupings  (Bootstrap,  Ultrafast  Bootstrap  Approximation  and  probabilistic  contrasts). The program can be downloaded and run locally, or on online servers such as:
http://iqtree.cibiv.univie.ac.at/
https://www.phylo.org/
https://www.hiv.lanl.gov/content/sequence/IQTREE/iqtree.html
You can find many basic and advanced tutorials in http://www.iqtree.org/doc/


In the terminal, make sure you're still in the correct directory.

Type: `iqtree2 -h` to see all the command options for IQTREE, and note any that may be useful.

You may notice the `-m` argument to specify a model. You can run `-m TEST` when making a tree to find the best substitution model and then make a tree using that model. Alternatively, you can use `-m TESTONLY` to find the best fitting substitution model without making a tree.

Other important arguments we'll be using include:

`-s` to specify the name of the alignment file,always required by IQ-TREE to work.

`-B` to specify the number of replicates for ultrafast bootstrap in IQ-TREE v2. Slightly less accurate than regular bootstrapping, but MUCH faster.

`-b` to specify the number of replicates for regular bootstrapping. Either this or `-B` can be used, not both.


To start with, we'll run the modeltest without running a tree.

### Task 2
In the terminal, type `iqtree -s omicron_subset_countries_aligned.fasta -m TESTONLY` 

What is the best fitting model?

Hint: This will be on the terminal screen, you may have to scroll up a little to under where each model was tested.

### Task 3
What do the parameters of the best fitting model mean?

## 4.2.1b.4: Tree building

Now we know the best fitting model, we can make our tree!

### Task 4

In the terminal write `iqtree -s omicron_subset_countries_aligned.fasta -m BEST_FITTING_MODEL -B 1000 -redo` replacing `BEST_FITTING_MODEL` with the model you found in your last step.

This command means we are building this tree with 1000 ultrafast bootstrap replicates. The `-redo` argument is necessary to overwrite the files made in the last step and allow this to run.

You'll notice a lot of new files appear in your directory. The ones we're most interested in are the `.contree` file, which we'll use to visualise our tree, and the `.log` file which contains all the information that appeared on the terminal window.


## 4.2.1b.5: Tree visualisation

Now you've built a tree, it's time to visualise it. We'll be using FigTree. FigTree is designed as a graphical viewer of phylogenetic trees and as a program for producing publication-ready figures. It can be downloaded from: http://tree.bio.ed.ac.uk/software/figtree/ 

### Task 5

Type `figtree` in the terminal to open the program.

Click  File → Open → select the file `omicron_subset_countries_aligned.contree`

When prompted, enter `bootstrap` as the label name.

Your tree will appear on the screen! However, it needs a bit of manipulation to be in its most useful form.

Along the side tabs, go to Trees. Root the tree at the midpoint and order the nodes by increasing order.

### Task 6
What initial observations can you make from the tree topology?

### Task 7
Now go to Node labels, and display the bootstrap values. 

What is the purpose of bootstrap values?

### Task 8

It often isn't useful to see ALL the bootstrap values - we just want to know which branches have strong support and which don't. Therefore, we can hide the lower values and only show values of interest. 

To do this, click `ctrl+f`. A search bar will open at the top of the figure. Search for bootstrap values of at least 90. Click `Node` in selection mode, then click the `annotate` paperclip. This will bring up a box. Select `bootstrap` from the drop down list, leave the second box blank, and apply. 

This will remove any bootstrap values less than 90. 

Look at the bootstrap values on the tree now. What do they tell us?

### Task 9

Spend some time now playing about with the settings in FigTree to make an aesthetically pleasing tree! Make sure to keep it open when you're done, as we'll be going back to this.


## 4.2.1b.6: Extracting metadata

We can add additional information to our tree. Sometimes the sequence IDs contain useful information that we want to extract. 

### Task 10

Take a look at the format of the sequence IDs in your fasta file.

What information might we want to extract from the sequence IDs?

### Data Processing in R

To extract this information manually would take ages! Instead, we can write a pipeline in R to process this for us.

### Task 11

Open RStudio using https://vm1.mbldash.tk/ (this may differ for different VMs - the link will be given in the session). Sign in using your assigned user as your username and password e.g. user1

You'll then need to run the command `.libPaths("/home/manager/R/x86_64-pc-linux-gnu-library/4.2")` to source the relevant R libraries

### Task 12

You'll need to import your fasta file. This first involves knowing where it is. You can find this out using the command `file.choose()`. This will open a window where you can select your alignment file. The full path to this will appear in the console in R studio. 

To import and store your alignment in R, run:
`alignment<-seqinr::read.alignment("PATH_TO_FILE", format = "fasta")` replacing the PATH_TO_FILE with the path you generated in the last step. 

Running this command should result in an element called `alignment` appearing in the environment tab in Rstudio. 

### Task 13
We're now going to create a table listing the country for each sequence. First, we'll create a table with all the sequence IDs and an empty country column ready to fill with data:

`m<-data.frame(ID = alignment$nam, country = NA)` 

This table should now have appeared under the alignment in your environment tab! 

### Task 14

You should have seen from looking at the sequence IDs that the information in the ID is separated by `/`s, with the country being the first element. We can therefore write some code to split the ID by the `/` and extract the first element:

`m$country<-sapply(strsplit(m$ID, "/"), "[", 1)` 

If you click on table `m` in your environment tab, you should now see that the country column has been filled in! 

### Task 15

We now want to save this table in a form that can be used by FigTree to annotate our tree. This is tab delimited text. To save this:

`write.table(m, file = "PATH_TO_DIRECTORY/omicron_subset_countries.txt", sep = "\t", row.names = FALSE)` 

Replacing the PATH_TO_DIRECTORY with the full file path to your analysis directory (one step up from the file path we generated last time!).

If we check in our directory, we should now see the `.txt` file!

## 4.2.1b.7: Annotating the tree

If we go back to our FigTree tree, we can now annotate it with the metadata file we just created! We do this by going to File->Import annotations and selecting the txt file we just made. 

We can now colour our tips by country! In the side tab, go to Tip labels and colour by country.

### Task 16

Again, play around until you have a tree you're happy with. 

When you've finalised your tree, you can export it by File->Export X where X is the file type. 

Alternatively, you can try to annotate by different metadata!

## 4.2.1b.7: Additional activity: Extracting alternate metadata

Go back to RStudio and try to extract some other useful metadata!

First, try extracting the year for each sequence. 

Most of the code will be the same as you ran before, but think about the element of the ID you want to extract! Hint: The working code will be at the end of this document if you need assistance!

### Task 17

Annotate your tree in FigTree with tips coloured by year

It might also be useful to colour tips by country of interest - e.g. colour sequences from the Philippines one colour, and other countries another colour. 

You could do this by editing the country `.txt` file you made manually in excel. Or we can practice some R skills!

There are a few ways of doing this, but to save writing additional code, we can start with the same code we ran before to extract the countries.

Before we save the table, we want to convert any country that doesn't say `Philippines` to `Other`.

To do this, we need to find which entries do not say `Philippines`. For this we use the `!=` operator, meaning 'not equal to'. We then need to tell R which positions these entries are in the table (their row numbers), and tell it to change these row entries to 'Other':

`m$country[which(m$country != "Philippines")]<-"Other"`

Check table `m` to make sure this has changed, then save the file!

### Task 18

Annotate your tree in FigTree with tips colourd by PHL/Other

Code for year colouring: 

`m<-data.frame(ID = alignment$nam, year = NA)`
`m$year<-sapply(strsplit(m$ID, "/"), "[", 3)`
